### PR TITLE
Load quantized checkpoint weights per component

### PR DESCRIPTION
**Pull request by xiaoyu-work**
## Summary

- Add typed weight records, component-owned bundles, and the Olive/GPTQ/AWQ codec registry.
- Load existing quantized checkpoint weights using independent component layouts and per-projection HuggingFace source-name overrides and exclusions; do not quantize floating-point checkpoint weights.
- Preserve Gemma4, Qwen3.5/QMoE, and T5 architecture-specific weight adaptations, tied tables, and component sidecar routing, and validate quantized parameter binding.
- Preserve default attention head dimensions and KV-head counts when HuggingFace configuration overrides only a subset of layers.
- Keep Gemma4 per-layer inputs in `[batch, sequence, layers, projection]` form between embedding and decoder components for OpenVINO; other execution providers retain the flattened layout.
- Restore the seq2seq, vision-language, and speech independent-component quantization matrix coverage lost during the earlier graph-test file split.

## Loading correctness fixes

- Recognize canonical weights only as complete, shape-matching groups so raw GPTQ/AWQ scales remain available to their codec; validate zero points using each projection's effective symmetry.
- Normalize quantized embedding tables for two-dimensional GatherBlockQuantized storage rather than treating them as linear projections.
- Honor Qwen3.5-MoE's model-declared float exclusions and Whisper's declared output head during both graph construction and loading.
- Preserve the distinction between an absent component mapping (legacy global rules) and an explicitly empty mapping (float components).
- Defer Gemma4 component packing until projection overrides are resolved. Restore T5 HuggingFace source aliases and separate decoder-stack rules from the independently owned top-level LM head.

## Consolidation

Replaces #666, #669, and #670 and consolidates the related development from `test/per-component-quantization-stack` into this single PR.

The branch starts directly from `main` at `a471d3416e9c52d5508d28730962e18c8a1053d0`. The design (#664) and canonical component manifests (#665) are already merged and are not separate dependencies of this PR.

This includes all seven previously uncommitted Gemma4/config/EP files from the old development worktree, not just the three original PR heads. The committed old feature stack was also compared with main plus this PR: production feature areas were retained, and its missing graph-layout matrix tests were restored in the current split test modules.

## Review and CI follow-up

Addressed all eight inline review threads, including duplicate/already-resolved reports, in `190eb8a5`:

- Literal module rules now respect path boundaries; the compatibility normalizer rejects a missing task clearly. Redundant Protocol statements and alias parentheses are removed.
- Fixed the Gemma3 split-tied loading regression, including its previously floating-point scaled embedding. A shared scaled quantized embedding preserves Gemma4 behavior, and actual embedding execution plus unsupported-tie guards are covered.
- Corrected the LLaVA fixture's missing `vision_model` namespace without relaxing target validation.
- Updated the stale GGUF integration expectation to the export contract already established by #688; tokenizer-identity rejection and full logit/generation comparisons remain intact.

The original full L4/L5 jobs hit their explicit one-hour limits after recording failures. Those failed case IDs also occur on the exact base `a471d341`, but cancellation lost exception details, so they are not claimed to be infrastructure-only or fixed. No model selection, tolerance, or timeout was weakened. Isolated GPT-2 diagnostics at the original head each executed and passed exactly one selected case: L4 https://github.com/onnxruntime/mobius/actions/runs/34308638240 and L5 https://github.com/onnxruntime/mobius/actions/runs/34308638242 . These single-case successes do not certify the all-model golden jobs.

Fresh full CI for the published follow-up: https://github.com/onnxruntime/mobius/actions/runs/34309655380 .

## Validation

- **874 passed, 16 skipped, 4 deselected** in the focused loading/config/model/CLI/graph suite, including the repaired CI selectors and scaled quantized embedding execution.
- **1 passed** for the real pinned `smollm2-135m-instruct-f16` GGUF CLI export/reload, full-logit, and generation integration case.
- Ruff lint and formatting passed for the changed Python files. Independent review of this follow-up found no significant issues.
- Runs used the repository-required onnx-ir API through isolated dependencies; the shared environment was not modified.
- The original seven-file pending patch remains consolidated in this PR. Local evidence does not claim that the all-model golden jobs or physical NPU execution are green.

### CODE REVIEW

File: src/mobius/weights/_adapters.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    """Architecture-specific rename/split/fuse operations only."""
+
+    def adapt(
+        self,
+        module: nn.Module,
+        state_dict: Mapping[str, torch.Tensor],
+        context: WeightAdapterContext,
+    ) -> dict[str, torch.Tensor]:
+        """Return semantically aligned weights without format normalization."""
+        ...
```
[the remark this answers is not in the document]
Comment (xiaoyu-work) [reply src/mobius/weights/_adapters.py]: Removed the redundant ellipsis from this Protocol method in 190eb8a5. The docstring remains the stub body; this is a cleanup only.

File: src/mobius/weights/_codecs.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        """Group checkpoint sidecars into typed logical records."""
+        ...
+
+    def normalize(
+        self,
+        record: WeightRecord,
+        config: QuantizationConfig,
+    ) -> dict[str, torch.Tensor]:
+        """Convert one packed record to Mobius's canonical parameter layout."""
+        ...
```
[the remark this answers is not in the document]
Comment (xiaoyu-work) [reply src/mobius/weights/_codecs.py]: Removed the redundant ellipsis from this Protocol method in 190eb8a5. The docstring remains the stub body; the codec contract is unchanged.

File: src/mobius/_component_quantization.py [hunk 3]
```diff
@@ -336,399 +162,557 @@ def _embedding_layout_matches(
...
+
+def preprocess_component_quantized_state_dict(
+    state_dict: dict[str, torch.Tensor],
+    module: nn.Module,
+    config: BaseModelConfig,
+    task: ModelTask | str | None,
+    package_components: Iterable[str],
+) -> dict[str, torch.Tensor]:
+    """Compatibility wrapper for normalize_component_quantized_weights."""
+    resolved_task = get_task(task) if task is not None else None
```
[the remark this answers is not in the document]
Comment (xiaoyu-work) [reply src/mobius/_component_quantization.py]: Addressed in 190eb8a5. The compatibility normalizer now requires an explicit task and raises a clear ValueError when None is passed, before manifest resolution.

File: src/mobius/_configs/_quantization.py [hunk 4]
```diff
@@ -220,145 +225,52 @@ def from_transformers(cls, hf_config) -> QuantizationConfig | None:
...
-        """Whether Olive recorded component-selection metadata."""
-        return self.modules_to_not_convert is not None or bool(self.overrides)
+        """Whether module-selection rules or overrides are present."""
+        return bool(self.modules_to_not_convert) or bool(self.overrides)
+
+    @staticmethod
+    def _matches_exclusion(pattern: str, module_name: str) -> bool:
+        if pattern.startswith("re:"):
+            return _compile_pattern(pattern[3:]).fullmatch(module_name) is not None
+        return pattern in module_name
```
[the remark this answers is not in the document]
Comment (xiaoyu-work) [reply src/mobius/_configs/_quantization.py]: Addressed in 190eb8a5. Literal rules now match the exact HuggingFace module path or dot-separated descendants, so layers.1 does not match layers.10, layers.1_extra, or an unrelated prefix. Regex rules still use full matching.

File: src/mobius/models/gemma4.py [hunk 5]
```diff
@@ -3471,6 +3486,25 @@ class Gemma4Model(nn.Module):
...
+        },
+        "audio_encoder": {
+            "encoder": "model.audio_tower",
+            "projector": "model.embed_audio.embedding_projection",
+        },
+        "embedding": {
+            "embed_tokens": "model.language_model.embed_tokens",
+            "embed_tokens_per_layer": ("model.language_model.embed_tokens_per_layer"),
+            "per_layer_model_projection": ("model.language_model.per_layer_model_projection"),
+        },
```
[the remark this answers is not in the document]
Comment (xiaoyu-work) [reply src/mobius/models/gemma4.py:3426]: Simplified in 190eb8a5: removed the redundant parentheses around the embedding alias strings. The mapping values and behavior are unchanged.

File: src/mobius/weights/_codecs.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+    method: str
+
+    def group(
+        self,
+        component: ComponentDescriptor,
+        state_dict: Mapping[str, torch.Tensor],
+        config: QuantizationConfig,
+    ) -> WeightBundle:
+        """Group checkpoint sidecars into typed logical records."""
+        ...
```
[the remark this answers is not in the document]
Comment (xiaoyu-work) [reply src/mobius/weights/_codecs.py]: Removed the redundant ellipsis from this Protocol method in 190eb8a5. The docstring remains the stub body; the codec contract is unchanged.

File: src/mobius/_component_manifest.py [hunk 7]
```diff
@@ -158,13 +165,18 @@ def resolve_component_manifest(
...
+            model_type or "",
             hf_config,
         )
-        raw_aliases = getattr(module_class, "HF_COMPONENT_MODULE_ALIASES", {})
+        alias_resolver = getattr(module_class, "get_hf_component_module_aliases", None)
+        raw_aliases = (
+            alias_resolver(model_type=model_type or "", hf_config=hf_config)
+            if alias_resolver is not None
+            else getattr(module_class, "HF_COMPONENT_MODULE_ALIASES", {})
+        )
```
[the remark this answers is not in the document]
Comment (xiaoyu-work) [reply src/mobius/_component_manifest.py:182]: Addressed in ad666a82. Dynamic source hooks now use the explicit model type or hf_config.model_type; when neither exists, they are not invoked and no static paths are guessed for classes with dynamic source hooks. Static metadata remains available for classes without dynamic hooks. Config-based alias hooks now take only hf_config (the old model_type argument was unused by T5), preserving hand-built T5 aliases without passing a fabricated empty type. No quantization policy or packing behavior was changed.

**Review by titaiwangms — COMMENTED** [review]
Full-team review of `ad666a825b9becf297ae350b5a98f05dc2ae831a` (readability, correctness, adversarial, semantic-invariant, and integration passes).

The typed records/codec boundary and final binding check are strong improvements, but I found four correctness issues that should be addressed before merge:

1. **Every plain `Embedding` in a quantized component is now treated as a token table.**  
   `src/mobius/_component_quantization.py:346-366` rewrites any ordinary `Embedding` when `quantize_embeddings=True`. The previous implementation restricted this to known token-table names, and this PR removes the test that ensured positional embeddings remained float. Components such as speech decoders can own both token and learned positional embeddings; a positional table may therefore become `GatherBlockQuantized` even though the checkpoint contains only its float weight, leaving packed inputs unbound. Please make this decision manifest/source-name driven (or otherwise restore a token-table-only gate) and add a non-token embedding regression test.

2. **Qwen3.5 component adapters still normalize ordinary decoder sidecars before per-projection overrides are resolved.**  
   `src/mobius/models/qwen35.py:582-588`, `:702-722`, and `:988-1005` call `preprocess_quantized_weights` with a decoder-wide configuration. `QuantizationConfig.for_source_paths()` (`src/mobius/_configs/_quantization.py:256-266`) resolves the source roots and strips the remaining module rules. A leaf override such as `model.language_model.layers.0.self_attn.q_proj` can therefore build an overridden graph through the generic per-module rewrite while the adapter has already packed its weight with the component default. The shared loader sees canonical keys and cannot correct the layout. Please keep non-expert sidecars raw in component mode, as Gemma4 does, and limit the adapter to rename/tie/QMoE-specific transformations.

3. **The graph and weight paths disagree on when legacy global module rules are active.**  
   `configure_component_quantization()` gates them on `manifest.names == ("model",)` (`src/mobius/_component_quantization.py:428-434`), while `normalize_component_quantized_weights()` gates them on `len(component_names) == 1` (`:610-616`). Single-component tasks named `decoder`, `vision_encoder`, etc. can consequently normalize with exclusions/overrides that were never applied to graph construction. Please use one shared predicate based on the same manifest/package ownership contract.

4. **Binding validation skips that newly supported legacy module-rule path.**  
   `validate_quantized_component_bindings()` returns whenever `component_quantization is None` (`src/mobius/_component_quantization.py:740-746`), even though configure/normalize explicitly support a root `quantization` containing exclusions or overrides. Missing packed weights/scales can therefore survive the loader and fail only during session creation or inference. The validator can safely scan emitted quantized ops regardless of whether the plan came from the component mapping or the root config.

Additional issues:

- `normalize_component_quantized_weights()` skips codec grouping when there is no raw qweight (`src/mobius/_component_quantization.py:628-632`), and its final residual check covers only qweight keys (`:728-730`). An unrouted/orphan `*_scales`, `*_qzeros`, `.scales`, or `.qzeros` entry is silently accepted. Validate all packed sidecar suffixes and add a scales-only regression test.
- The design contract says `QuantizationCodec.normalize(...) -> WeightRecord`, while the implemented protocol returns `dict[str, torch.Tensor]` (`docs/design/per-component-quantization-loading.md`, codec section vs. `src/mobius/weights/_codecs.py:174-180`). Please update the design snippet so future codecs implement the actual interface.

Positive notes: complete canonical-group detection is substantially safer than suffix-only detection; zero-point validation occurs after effective override resolution; and CLI/Transformers now share the same manifest → adapter → codec → binding pipeline.

**Review by justinchuby — APPROVED** [review]

**Comment by xiaoyu-work** [comment]
@copilot resolve the merge conflicts in this pull request

**Comment by xiaoyu-work** [comment]
@copilot resolve the merge conflicts in this pull request

**Review by titaiwangms — APPROVED** [review]

**Comment by xiaoyu-work** [comment]
@copilot resolve the merge conflicts in this pull request
